### ISSUE TITLE
Categories and assigned bars getting mixed up

### ISSUE THREAD

**Issue by hollo6** [arkayenro/arkinventory#837]
Frequently (most of the time, like in 80% of the cases), items land in places in the bag where they don't belong. Sometimes they seem to forget their assigned category (custom, changed built-in and default as well), sometimes they have the category (according to the tooltip) but they are not in their assigned bar. Refresh button does not help. Reload UI helps in ~20% of the cases, meaning if I reload like 5-6 times, items eventually land in their intended places.
I'm using 3.07.05. I think I have noticed this with 3.07.04 but not that frequently.

**Comment by hollo6** [comment]
After a few hours of messing around with reload ui and enabling/disabling some other addons, it seems this is caused by a race condition on load. Would it be possible to check if the bag category assignments have all been loaded properly at the end? (like that translations check) My experience so far is that most of the collisions are with TSM (the base addon, it does this if no modules are enabled and enabling any modules doesn't affect the collision rate).

**Comment by arkayenro** [comment]
in ArkInventory.toc theres a line that starts with ## OptionalDeps

add <comma>TradeSkillMaster to the end of it and see if it fixes it or not.  that should force TSM to load before AI does if you have it enabled

**Comment by hollo6** [comment]
This helped a little. Now bag categories load successfully every time, but bank is still hectic, mostly wrong.
I don't know if this has something to do with the issue but the bank problem mostly affects equippable items (ones that have custom category or rule).
I was just guessing that race condition because which categories fail to load seem to be random.

**Comment by hollo6** [comment]
New discovery: entering edit mode makes the items in the bank go to their intended places.

**Comment by kerbaal** [comment]
I have this problem, too. I just added TradeSkillMaster to the OptionalDeps line and will report back.

Thanks for brining this issue up here, it's been annoying.

**Comment by arkayenro** [comment]
the tsm optdep change may only help people with tsm based rules, it wouldnt help any other issue

what sorting do you have set?  instant, on open or manual?

restacking can move items around, the blizzard cleanup function moves items around a lot the first time its run - as above your sorting type will determine when the items are moved to their category locations

if you unequip the item what category is it assgined (check in edit mode) - also check its category in the wearing location (when its equipped)

**Comment by arkayenro** [comment]
> New discovery: entering edit mode makes the items in the bank go to their intended places.

edit mode just does a full resort, like clicking on refresh

> edit: not always. seems this "fix" requires a reload first or some time spent after login.

unless TSM is doing something odd?  there is no longer a way to check if TSM is ready, or if you change anything in TSM - so you have to do a /reload after any TSM changes for AI to pick them up

are you making changes in TSM?

**Comment by hollo6** [comment]
> > edit mode just does a full resort, like clicking on refresh
I don't think that is the only thing it does. If the mixup is present, refresh button does nothing, edit mode sometimes gets it fixed (if not, /reload and then edit mode does it).

I don't use any TSM rules and I'm not doing anything when testing these.

**Comment by hollo6** [comment]
Now I'm positive this problem has something to do with refreshing. Just found out that changing something in the rules (tested with changing order number) _always_ fixes it.

**Comment by kerbaal** [comment]
Changing a rule does not fix it for me. Changing assignments also does not fix it for me.

I have sorting method on open set. Opening and closing does not fix the problem. The funny thing is: after the ArkInventory window has been closed for about 30 seconds, opening it fixes the problem.

On my current setup I can reproduce the problem as follows.

1. Ensure Light Parchment assigned to Professions/Inscription (this might be the default).
2. Assign Professions/Inscription to a bag different from the default bag
2. Sell all Light Parchment, and close ArkInventory
3. Open ArkInventory and buy 5 Light Parchment. On my setup they are in the default bag now.
4. Open and Close ArkInventory.
5. Observe that Light Parchment is _still_ in the default bag, not in the bag Professions/Inscription was assigned to.

**Comment by hollo6** [comment]
Seems we have a different problem. I have also noticed that waiting sometimes helps though, but not always and I have to wait much longer (like 5-10 minutes).
Have you tried disabling TSM? Any of my problems don't appear if TSM is disabled.

**Comment by kerbaal** [comment]
I don't have TSM installed.

**Comment by kerbaal** [comment]
I got this error, maybe it is related:

```
Message: Interface\AddOns\ArkInventory\ArkInventory.lua:4352: script ran too long
Time: 09/02/16 01:04:19
Count: 1
Stack: Interface\AddOns\ArkInventory\ArkInventory.lua:4352: in function `Frame_Main_DrawThreadResume'
[string "*:OnUpdate"]:4: in function <[string "*:OnUpdate"]:1>

Locals: (for generator) = <function> defined =[C]:-1
(for state) = <table> {
 1 = <no value>
}
(for control) = 1
loc_id = 1
thread = <no value>
(*temporary) = <function> defined =[C]:-1
(*temporary) = <no value>
(*temporary) = "script ran too long"
pairs = <function> defined =[C]:-1
type = <function> defined =[C]:-1

```

**Comment by arkayenro** [comment]
in the workarounds section, lower the number of items before yielding
